lib/sources: remove 2 usages of toString on a path which will be read using fileContents

Artturin opened the pull request:
Check commit messages

###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Artturin committed:
lib/trivial: fix 'error: cannot decode virtual path '/nix/store/virtual0000000000000000000000005-source''

happens on lazy-trees branch of Nix (NixOS/nix#6530)

Artturin committed:
lib/sources: remove 2 usages of toString on a path which will be read using fileContents

It gives a warning on the lazy-trees branch of Nix
(NixOS/nix#6530)

"warning: applying 'toString' to path '...' and then accessing it is deprecated, at '...'"

'else toString (/. + "${base}/${path}");' at line 183 may still cause a warning but i don't know how
to reach that codepath and test so im leaving it untouched

changing it to 'else /. + "${base}/${path}";'
caused this error
```
error: a string that refers to a store path cannot be appended to a path

       at /home/systems/nixpkgs/lib/sources.nix:183:20:

          182|               then path
          183|               else /. + "${base}/${path}";
             |                    ^
          184|         in if pathIsRegularFile path
```

Artturin commented:
Getting the following in `nix flake check`

```
nixpkgs-tarball> checking eval-release.nix
nixpkgs-tarball> trace: lib.crossLists is deprecated, use lib.cartesianProductOfSets instead
nixpkgs-tarball> trace: warning: literalExample is deprecated, use literalExpression instead, or use literalDocBook for a non-Nix description.
nixpkgs-tarball> trace: warning: lib.mkFixStrictness has no effect and will be removed. It returns its argument unmodified, so you can just remove any calls.
nixpkgs-tarball> trace: `lib.nixpkgsVersion` is deprecated, use `lib.version` instead!
nixpkgs-tarball> trace: warning: lib.readPathsFromFile is deprecated, use a list instead
nixpkgs-tarball> trace: Warning: `showVal` is deprecated and will be removed in the next release, please use `traceSeqN`
nixpkgs-tarball> trace: lib.zip is deprecated, use lib.zipAttrsWith instead
nixpkgs-tarball> error: getting status of '/build/source/.git': No such file or directory
nixpkgs-tarball>        … while evaluating 'readCommitFromFile'
nixpkgs-tarball>        at /build/source/lib/sources.nix:177:36:
nixpkgs-tarball>           176|   commitIdFromGitRepo =
nixpkgs-tarball>           177|     let readCommitFromFile = file: path:
nixpkgs-tarball>              |                                    ^
nixpkgs-tarball>           178|         let fileName       = path + "/" + file;
nixpkgs-tarball>        … from call site
nixpkgs-tarball>        at /build/source/lib/sources.nix:169:35:
nixpkgs-tarball>           168|
nixpkgs-tarball>           169|   pathIsGitRepo = path: (tryEval (commitIdFromGitRepo path)).success;
nixpkgs-tarball>              |                                   ^
nixpkgs-tarball>           170|
nixpkgs-tarball>        … while evaluating 'pathIsGitRepo'
nixpkgs-tarball>        at /build/source/lib/sources.nix:169:19:
nixpkgs-tarball>           168|
nixpkgs-tarball>           169|   pathIsGitRepo = path: (tryEval (commitIdFromGitRepo path)).success;
nixpkgs-tarball>              |                   ^
nixpkgs-tarball>           170|
nixpkgs-tarball>        … from call site
nixpkgs-tarball>        at /build/source/lib/trivial.nix:218:11:
nixpkgs-tarball>           217|       gitRepo      = ./.. + "/.git";
nixpkgs-tarball>           218|     in if lib.pathIsGitRepo gitRepo
nixpkgs-tarball>              |           ^
nixpkgs-tarball>           219|        then lib.commitIdFromGitRepo gitRepo
nixpkgs-tarball>        … while evaluating 'revisionWithDefault'
nixpkgs-tarball>        at /build/source/lib/trivial.nix:214:5:
nixpkgs-tarball>           213|     # Default value to return if revision can not be determined
nixpkgs-tarball>           214|     default:
nixpkgs-tarball>              |     ^
nixpkgs-tarball>           215|     let
nixpkgs-tarball>        … from call site
nixpkgs-tarball>        at /build/source/doc/doc-support/lib-function-locations.nix:3:14:
nixpkgs-tarball>             2| let
nixpkgs-tarball>             3|   revision = pkgs.lib.trivial.revisionWithDefault (nixpkgs.revision or "master");
nixpkgs-tarball>              |              ^
nixpkgs-tarball>             4|
nixpkgs-tarball>        … while evaluating anonymous lambda
nixpkgs-tarball>        at /build/source/doc/doc-support/lib-function-locations.nix:62:8:
nixpkgs-tarball>            61|   xmlstrings = (nixpkgsLib.strings.concatMapStrings
nixpkgs-tarball>            62|       ({ name, value }:
nixpkgs-tarball>              |        ^
nixpkgs-tarball>            63|       ''
nixpkgs-tarball>        … from call site
nixpkgs-tarball>        … while evaluating 'concatMapStrings'
nixpkgs-tarball>        at /build/source/lib/strings.nix:54:25:
nixpkgs-tarball>            53|   */
nixpkgs-tarball>            54|   concatMapStrings = f: list: concatStrings (map f list);
nixpkgs-tarball>              |                         ^
nixpkgs-tarball>            55|
nixpkgs-tarball>        … from call site
nixpkgs-tarball>        at /build/source/doc/doc-support/lib-function-locations.nix:61:17:
nixpkgs-tarball>            60|   urlPrefix = "https://github.com/NixOS/nixpkgs/blob/${revision}";
nixpkgs-tarball>            61|   xmlstrings = (nixpkgsLib.strings.concatMapStrings
nixpkgs-tarball>              |                 ^
nixpkgs-tarball>            62|       ({ name, value }:
nixpkgs-tarball>        … while evaluating the attribute 'text' of the derivation 'locations.xml'
nixpkgs-tarball>        at /build/source/pkgs/stdenv/generic/make-derivation.nix:270:7:
nixpkgs-tarball>           269|     // (lib.optionalAttrs (attrs ? name || (attrs ? pname && attrs ? version)) {
nixpkgs-tarball>           270|       name =
nixpkgs-tarball>              |       ^
nixpkgs-tarball>           271|         let
nixpkgs-tarball>        … while evaluating the attribute 'buildCommand' of the derivation 'doc-support'
nixpkgs-tarball>        at /build/source/pkgs/stdenv/generic/make-derivation.nix:270:7:
nixpkgs-tarball>           269|     // (lib.optionalAttrs (attrs ? name || (attrs ? pname && attrs ? version)) {
nixpkgs-tarball>           270|       name =
nixpkgs-tarball>              |       ^
nixpkgs-tarball>           271|         let
nixpkgs-tarball>        … while evaluating the attribute 'postPatch' of the derivation 'nixpkgs-manual'
nixpkgs-tarball>        at /build/source/pkgs/stdenv/generic/make-derivation.nix:270:7:
nixpkgs-tarball>           269|     // (lib.optionalAttrs (attrs ? name || (attrs ? pname && attrs ? version)) {
nixpkgs-tarball>           270|       name =
nixpkgs-tarball>              |       ^
nixpkgs-tarball>           271|         let
nixpkgs-tarball>        … while evaluating the attribute 'drvPath'
nixpkgs-tarball>        at /build/source/lib/customisation.nix:210:7:
nixpkgs-tarball>           209|     in commonAttrs // {
nixpkgs-tarball>           210|       drvPath = assert condition; drv.drvPath;
nixpkgs-tarball>              |       ^
nixpkgs-tarball>           211|       outPath = assert condition; drv.outPath;
nixpkgs-tarball>        … while evaluating 'recurse'
nixpkgs-tarball>        at /build/source/maintainers/scripts/eval-release.nix:13:19:
nixpkgs-tarball>            12|   # nix-instantiate recurses into nested attribute sets.
nixpkgs-tarball>            13|   recurse = path: attrs:
nixpkgs-tarball>              |                   ^
nixpkgs-tarball>            14|     if (builtins.tryEval attrs).success then
nixpkgs-tarball>        … from call site
nixpkgs-tarball>        at /build/source/maintainers/scripts/eval-release.nix:21:69:
nixpkgs-tarball>            20|       else { recurseForDerivations = true; } //
nixpkgs-tarball>            21|            mapAttrs (n: v: let path' = path ++ [n]; in trace path' (recurse path' v)) attrs
nixpkgs-tarball>              |                                                                     ^
nixpkgs-tarball>            22|     else { };
nixpkgs-tarball>        … while evaluating anonymous lambda
nixpkgs-tarball>        at /build/source/maintainers/scripts/eval-release.nix:7:76:
nixpkgs-tarball>             6| let
nixpkgs-tarball>             7|   trace = if builtins.getEnv "VERBOSE" == "1" then builtins.trace else (x: y: y);
nixpkgs-tarball>              |                                                                            ^
nixpkgs-tarball>             8|
nixpkgs-tarball>        … from call site
nixpkgs-tarball>        at /build/source/maintainers/scripts/eval-release.nix:21:56:
nixpkgs-tarball>            20|       else { recurseForDerivations = true; } //
nixpkgs-tarball>            21|            mapAttrs (n: v: let path' = path ++ [n]; in trace path' (recurse path' v)) attrs
nixpkgs-tarball>              |                                                        ^
nixpkgs-tarball>            22|     else { };
nixpkgs-tarball>        … while evaluating anonymous lambda
nixpkgs-tarball>        at /build/source/maintainers/scripts/eval-release.nix:21:25:
nixpkgs-tarball>            20|       else { recurseForDerivations = true; } //
nixpkgs-tarball>            21|            mapAttrs (n: v: let path' = path ++ [n]; in trace path' (recurse path' v)) attrs
nixpkgs-tarball>              |                         ^
nixpkgs-tarball>            22|     else { };
nixpkgs-tarball>        … from call site
nixpkgs-tarball>        … while evaluating the attribute 'manual'
nixpkgs-tarball> build times:
nixpkgs-tarball> user time for the shell             0m0.000s
nixpkgs-tarball> system time for the shell           0m0.000s
nixpkgs-tarball> user time for all child processes   0m0.000s
nixpkgs-tarball> system time for all child processes 0m0.000s
```

remember to use https://github.com/NixOS/nix/pull/6530

roberth commented on lib/sources.nix:
```diff
@@ -175,8 +175,8 @@ let
     let readCommitFromFile = file: path:
-        let fileName       = toString path + "/" + file;
-            packedRefsName = toString path + "/packed-refs";
+        let fileName       = path + "/" + file;
```
```suggestion
        let fileName       = path + "/${file}";
```

`+` is left associative and `path + "/"` is a no-op.

Artturin replied:
Thanks

Artturin commented:
> Getting the following in `nix flake check` / `nix build -f ./pkgs/top-level/release.nix tarball`
> ... 
> remember to use [NixOS/nix#6530](https://github.com/NixOS/nix/pull/6530)

those commands work on master

roberth committed:
lib/sources: Make pathIsGitRepo not evaluate toString path

This requires us to avoid the `tryEval` + `throw` combination,
because throw is strict in its error message, and we don't want
to drop our single clue when `commitIdFromGitRepo` is used
incorrectly.

roberth commented:
The remaining problem was the `pathIsGitRepo` was using `tryEval`, but `throw` is strict in its argument, so it failed to `+ path`, which should have been `+ toString path` anyway. I've taken the liberty to push a fix.

tobiasBora commented:
Is this PR supposed to solve the fact that `nixos-rebuild` takes a lot of RAM (~1.3G) when the input is a flake (not tried without flake) as it tries to evaluate the whole nixpkgs repository? Or is it unrelated?

Artturin commented:
> Is this PR supposed to solve the fact that `nixos-rebuild` takes a lot of RAM (~1.3G) when the input is a flake (not tried without flake) as it tries to evaluate the whole nixpkgs repository? Or is it unrelated?

unrelated but https://github.com/NixOS/nix/pull/6530 may help

tobiasBora commented:
Ok thanks @Artturin . For reference I created a separate issue there https://github.com/NixOS/nix/issues/7308

Artturin commented:
Thanks @roberth

zowoq commented:
https://github.com/NixOS/nixpkgs/actions/runs/3494613825/jobs/5850544462

This PR seems to have broken nixpkgs manual build.

vcunat commented:
Right, bisected to ec8f8f69bd57b in particular.  I don't understand why this happens, at least for now.  But if noone does very soon, we should revert.

vcunat commented:
Well, removing the doc-comment for `commitIdFromGitRepoOrError` makes the error go away, but I have no idea what's wrong about it and even simplifying to one simple sentence didn't help 🤯   (the XML errors here don't help me)

vcunat commented:
Apparently this is related to the combination of this new function not getting exported from the file, while still getting documented.  It's actually not good to get nixpkgs docs for a function that can't be used (from outside this file), so we can

A.  export it
```diff
--- a/lib/sources.nix
+++ b/lib/sources.nix
@@ -281,2 +281,3 @@
     commitIdFromGitRepo
+    commitIdFromGitRepoOrError
 
```

B.  or not include it in the docs.  It's nice to keep description though, so e.g. this experiment worked for me:
```nix
  # Get the commit id of a git repo.
  #
  # Returns `{ value = commitHash }` or `{ error = "... message ..." }`.
  #
  # Example: commitIdFromGitRepo <nixpkgs/.git>
```

maxeaubrey commented:
I cannot rebuild any of my systems since this PR was merged, had to revert to commit~1 to have working systems again. #201779 accurately represents my issue, but I'm unaware of my [system](github.com/maxeaubrey/systems) being a particularly esoteric configuration.

Shouldn't this be reverted? It's been 4 days and master is, at least to me, not possible to build:
```
error: a string that refers to a store path cannot be appended to a path

       at /home/systems/nixpkgs/lib/sources.nix:193:30:

          192|               then path
          193|               else toString (/. + "${base}/${path}");
             |                              ^
          194|         in if pathIsRegularFile path
(use '--show-trace' to show detailed location information)
```
(I'm not pasting a full trace because the issue afaik covers it)

Artturin commented:
https://github.com/NixOS/nixpkgs/pull/202370
